hazelcast/src/main/java/com/hazelcast/internal/util/collection/DefaultedMap.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util.collection;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * A map decorator that overrides {@link #get(Object)} to return a default value
+ * if the delegated map does not contain the specified key. It differs from
+ * {@link org.apache.commons.collections.map.DefaultedMap} by <ol>
+ * <li> forwarding operations that have a default implementation in {@link Map}
+ *      to the delegated map, which can increase performance if the delegated
+ *      map has a custom implementation, and
+ * <li> using {@link #getOrDefault} to implement the new {@code get} behavior.
+ */
+public class DefaultedMap<K, V> implements Map<K, V> {
+    private final Map<K, V> map;
+    private final V defaultValue;
+
+    public DefaultedMap(Map<K, V> map, V defaultValue) {
+        this.map = map;
+        this.defaultValue = defaultValue;
+    }
+
+    @Override
+    public V get(Object key) {
+        return map.getOrDefault(key, defaultValue);
```

Comment:
this implementation breaks contract for `Map.get`, in particular this part:

> Returns the value to which the specified key is mapped, or null if this map contains no mapping for the key.

To obey the contract we should have dedicated method, eg. `getOrDefault(Object key)`.

Reply:
I think this Map is a bit excessive, looking at usages, there's nothing special/implicit that would warrant this. I'd delete it and use normal HashMap.getOrDefault.

Is there a specific reason to have to have this Map impl beyond convenience for one particular use case?

Reply:
`DefaultedMap` doesn't stem from an unusual need. We have dozens of examples of the following pattern. In order to optimize those _resolve_ methods, we need static final variables to keep the mapping and potentially-complex default value. If we encapsulate the mapping, default value and convenience method, we obtain `DefaultedMap`.
```java
public static V resolveValue(K key) {
    switch (key) {
        case key1: return value1;
        case key2: return value2;
        // ...
        default: return defaultValue;
    }
}
```